### PR TITLE
Fixing the filter sizes

### LINKED ISSUE #37: Shape of the output of cnv.tl.infercnv
Hi all,

I noticed that when the window size is greater than the number of genes in a chromosome, the CNV profile for that part of the chromosome has the length of the window size and not the length of the original GEX. This is because the return length of np.convole with mode="same" is the maximum of the two input length.

This breaks the connection between CNVs and genes. Therefore, making it difficult to correctly aggregate CNV profiles from different patients.

Here is a minimal code example:

````
import anndata
import infercnvpy as cnv
import numpy as np

X = np.ones((4, 25))
adata = anndata.AnnData(X=X)

adata.obs["reference"] = ["reference"] * 2 + ["non-reference"] * 2
adata.var["chromosome"] = ["chr1"] * 25
adata.var["start"] = np.arange(25)
adata.var["end"] = np.arange(25)+1


cnv.tl.infercnv(adata,
                reference_cat="reference",
                reference_key="reference",
                window_size=50,
                step=1)


assert adata.obsm["X_cnv"].shape == adata.shape
````

If this is unintended behavior I am happy to provide a PR with a fix.

Best
Florian

### PR DESCRIPTION
**Pull request by redst4r**
Hi,
just came across two bugs in the `_running_mean()` function:
1. the size of the filter (`pyramid`) is off by one (`len(pyramid) == window_size-1`. Not a big deal, but slightly inconsistent.
2. as mentioned by #37, due to the way `np.convolve()` works, if the filtersize is bigger than the number of genes, it flips filter and signal, and the result doesn't mean much

1. is an easy fix
2. Two options: either skip the chromosome entirely if there's not enough genes. Or, make the filter smaller, essentially the chromosome yields a single convolution result (all genes on the chromosome go into the convolution). I opted for 2), especially for larger filter sizes, you'd loose alot of data when doing 1)

Thanks for putting together this nice package!

Closes #37

### CODE REVIEW

**Comment by grst** [comment]
Thanks for fixing this!

Do you urgently need this in a released version? Ideally, I would wait a couple of weeks until I fully ported infercnvpy to the [cookiecutter-scverse](https://github.com/scverse/cookiecutter-scverse) template, which should also fix the CI.

**Comment by redst4r** [comment]
Hi,
no need to put this into a release soon, I'll just merge it on my local repository for the time being!

File: src/infercnvpy/tl/_infercnv.py [hunk 1]
```diff
@@ -164,12 +164,21 @@ def _running_mean(x: Union[np.ndarray, scipy.sparse.spmatrix], n: int = 50, step
...
-    pyramid = np.minimum(r, r[::-1])
-    smoothed_x = np.apply_along_axis(lambda row: np.convolve(row, pyramid, mode="same"), axis=1, arr=x) / np.sum(
-        pyramid
-    )
-    return smoothed_x[:, np.arange(0, smoothed_x.shape[1], step)]
+    if n < x.shape[1]:  # regular convolution: the filter is smaller than the #genes
+        r = np.arange(1, n + 1)
+
+        pyramid = np.minimum(r, r[::-1])
+        smoothed_x = np.apply_along_axis(lambda row: np.convolve(row, pyramid, mode=conv_mode), axis=1, arr=x) / np.sum(
```
[the remark this answers is not in the document]
Comment (redst4r) [reply src/infercnvpy/tl/_infercnv.py]: ah, my mistake, I was playing around with the convolution mode in another branch and it leaked into here...

PS: is there a specific reason to have the convolution  with `mode="same"`? 
Seems to create weird edge effects on the end of chromosomes. The filter keeps sliding past the end of the chromosome, so the last datapoints at the start/end of the chromosome are convoluted with a "ramp" (half of the pyramid filter). Also, that makes the effective filter size vary along the chromosome (if `window_size=100` genes) you'll be smoothing over 100 genes most of the time, but at start and end of the chromosome, you'll just be smoothing 50 genes instead.
I've been using `mode="valid", that at least treats each position the same way.

I'm not sure what's more appropriate, maybe `mode=valid` makes detecting CNVs at the edges harder...

[the remark this answers is not in the document]
Comment (grst) [reply src/infercnvpy/tl/_infercnv.py]: Don't remember if there was a specific reason. Maybe I tried being consistent with the R version, but not sure if that's what they are doing. 

Personally, I don't think it matters that much, as it's just a very small region, and single-cell based CNV calling is anyway qualitative at best.

**Pull request merged**